### PR TITLE
Merge Runtime C API and C++ bindings

### PR DESCRIPTION
**Pull request by Wodann**
Something went wrong with https://github.com/mun-lang/runtime-ffi/pull/1 when force pushing that caused the PR to be automatically closed for good.

This PR includes the requested fixes for documentation.

### CODE REVIEW

File: include/mun/invoke_result.h [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+    private:
+        template <std::size_t... Is>
+        InvokeResult<Output, Args...> retry_impl(std::index_sequence<Is...>) {
+            auto err = std::move(unwrap_err());
+            auto& retry_fn = std::get<0>(err);
+            auto& update_fn = std::get<1>(err);
+            auto& args = std::get<2>(err);
+            while (!update_fn()) {
+                // Spin until an update has occured that might fix the error
```
Comment (baszalmstra) [inline include/mun/invoke_result.h]: Maybe at least put a std::this_thread::sleep in there? Or a yield?

File: include/mun/runtime.h [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+            auto arg2_guid = type_guid<Arg2>();
+            return std::equal(std::begin(arg1_guid.b), std::end(arg1_guid.b), std::begin(arg2_guid.b), std::end(arg2_guid.b));
+        }
+
+        template <typename Output, typename... Args>
+        constexpr bool valid_fn(const MunFunctionSignature& signature) noexcept {
+            constexpr auto NUM_ARGS = sizeof...(Args);
+
+            if (signature.num_arg_types != NUM_ARGS) {
+                std::cerr << "Invalid number of arguments. Expected: " << std::to_string(NUM_ARGS) << ". Found: " << std::to_string(signature.num_arg_types) << "." << std::endl;
```
Comment (baszalmstra) [inline include/mun/runtime.h]: This line is very long, maybe distribute over multiple lines?

File: include/mun/runtime.h [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+                return false;
+            }
+
+            if constexpr (NUM_ARGS > 0) {
+                const MunTypeInfo* arg_ptr = signature.arg_types;
+                const bool valid_args[] = { valid_arg<Args>(*(arg_ptr++))... };
+                for (size_t idx = 0; idx < NUM_ARGS; ++idx) {
+                    if (!valid_args[idx]) {
+                        const char* arg_names[] = { type_name<Args>()... };
+                        std::cerr << "Invalid argument type at index " << idx << ". Expected: " << signature.arg_types[idx].name << ". Found: " << arg_names[idx] << "." << std::endl;
```
Comment (baszalmstra) [inline include/mun/runtime.h]: Also very long

File: include/mun/runtime.h [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+                        return false;
+                    }
+                }
+            }
+
+            if (signature.return_type) {
+                if (!valid_arg<Output>(*signature.return_type)) {
+                    std::cerr << "Invalid return type. Expected: " << signature.return_type->name << ". Found: " << type_name<Output>() << "." << std::endl;
```
Comment (baszalmstra) [inline include/mun/runtime.h]: Same

File: include/mun/runtime.h [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+                    return false;
+                }
+            } else if (!equal_args<void, Output>()) {
+                std::cerr << "Invalid return type. Expected: " << type_name<void>() << ". Found: " << type_name<Output>() << "." << std::endl;
```
Comment (baszalmstra) [inline include/mun/runtime.h]: You get it.

File: include/mun/runtime.h [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+
+    template<>
+    constexpr const char* type_name<bool>() {
+        return "@core::bool";
+    }
+
+    template<>
+    constexpr MunGuid type_guid<double>() {
+        return MunGuid{
+            { 143, 142, 214, 135, 13, 131, 110, 33, 172, 233, 214, 19, 161, 21, 42, 130 }
```
Comment (baszalmstra) [inline include/mun/runtime.h]: I prefer to write this as hex values to indicate that they are bytes.

File: include/mun/runtime.h [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+         * \param other an rvalue reference to a runtime
+         */
+        Runtime(Runtime&& other) noexcept
+            : m_handle(other.m_handle) {
+            other.m_handle._0 = nullptr;
+        }
+
+        /** Destructs a runtime */
+        ~Runtime() noexcept {
+            mun_runtime_destroy(m_handle);
```
Comment (baszalmstra) [inline include/mun/runtime.h]: Wont this fail if the handle was moved?

File: include/mun/runtime.h [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+        }
+
+        /** Retrieves `FunctionInfo` from the runtime for the corresponding `fn_name`.
+         *
+         * \param fn_name the name of the desired function
+         * \param fn_info an output reference to an optional `MunFunctionInfo` struct
+         * \return a possible error
+         */
+        Error get_function_info(std::string_view fn_name, std::optional<MunFunctionInfo>& fn_info) noexcept {
```
Comment (baszalmstra) [inline include/mun/runtime.h]: Wouldnt it make more sense to flip the return value and output. 
```
std::optional<...> get_function_info(std::string_view fn_name, Error* err = nullptr)
```

File: include/mun/runtime.h [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+
+    /** Invokes the runtime function corresponding to `fn_name` with arguments `args`.
+     *
+     * \param runtime the runtime
+     * \param fn_name the name of the desired function
+     * \param args zero or more arguments to supply to the function invocation
+     * \return an invocation result
+     */
+    template <typename Output, typename... Args>
+    InvokeResult<Output, Args...> invoke_fn(Runtime& runtime, std::string_view fn_name, Args... args) noexcept {
```
Comment (baszalmstra) [inline include/mun/runtime.h]: Why is this a free function instead of a member of `Runtime`?

File: include/mun/runtime_capi.h [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+/**
+ * Destructs the error corresponding to `error_handle`.
+ */
+void mun_error_destroy(MunErrorHandle error_handle);
+
+/**
+ * Retrieves the error message corresponding to `error_handle`. If the `error_handle` exists, a
+ * valid `char` pointer is returned, otherwise a null-pointer is returned.
+ */
+char *mun_error_message(MunErrorHandle error_handle);
```
Comment (baszalmstra) [inline include/mun/runtime_capi.h]: Should this return a `const char*`?

File: include/mun/runtime.h (hunk 7 above)
Comment (Wodann) [reply include/mun/runtime.h]: `mun_runtime_destroy` nicely handles `nullptr`

File: include/mun/runtime.h (hunk 9 above)
Comment (Wodann) [reply include/mun/runtime.h]: The choice is inspired by this philosophy: https://www.drdobbs.com/cpp/how-non-member-functions-improve-encapsu/184401197

File: include/mun/runtime.h (hunk 8 above)
Comment (Wodann) [reply include/mun/runtime.h]: Fixed.

File: include/mun/invoke_result.h [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+     * the error state (i.e. callbacks and arguments) necessary to retry.
+     */
+    template <typename Output, typename... Args>
+    class InvokeResult {
+        using result_type = InvokeResult<Output, Args...>;
+        using success_type = Output;
+        using retry_fn_type = std::function<result_type(Args...)>;
+        using update_fn_type = std::function<bool()>;
+        using args_type = std::tuple<Args...>;
+        using err_type = std::tuple<retry_fn_type, update_fn_type, args_type>;
```
[the remark this answers is not in the document]
Comment (Wodann) [reply include/mun/invoke_result.h]: Changed.

File: include/mun/runtime.h (hunk 2 above)
Comment (Wodann) [reply include/mun/runtime.h]: Changed.

File: include/mun/runtime.h (hunk 3 above)
Comment (Wodann) [reply include/mun/runtime.h]: Changed.

File: include/mun/runtime.h (hunk 4 above)
Comment (Wodann) [reply include/mun/runtime.h]: Changed.

File: include/mun/invoke_result.h (hunk 1 above)
Comment (Wodann) [reply include/mun/invoke_result.h]: I've added `std::this_thread::sleep` for now and added a TODO for adding wait functionality inside the Rust Runtime API. This is non-trivial, so will have to wait for now.

File: include/mun/runtime.h (hunk 6 above)
Comment (Wodann) [reply include/mun/runtime.h]: Changed.

File: include/mun/runtime_capi.h (hunk 10 above)
Comment (Wodann) [reply include/mun/runtime_capi.h]: Fixed.

File: include/mun/runtime.h (hunk 5 above)
Comment (Wodann) [reply include/mun/runtime.h]: Changed.

File: include/mun/runtime.h [hunk 12]
```diff
@@ -0,0 +1,303 @@
...
+    };
+
+    /** Construct a new runtime that loads the library at `library_path` and its dependencies.
+     *
+     * On failure, the error is logged to the error log.
+     *
+     * \param library_path the path to a Mun library
+     * \return an optional runtime
+    .*/
+    std::optional<Runtime> make_runtime(std::string_view library_path) noexcept {
```
Comment (baszalmstra) [inline include/mun/runtime.h]: I would expect to have the same error handling here as in the runtime itself instead of logging to cerr.

Comment (Wodann) [reply include/mun/runtime.h]: Fixed.

**Pull request merged**